examples/hosts/hosts-sample/src/app.ts

```diff
@@ -0,0 +1,172 @@
+        hostUrl,
+        ordererUrl,
+        storageUrl,
+        tenantId,
+        bearerSecret,
+    );
+
+    const tokenProvider = new InsecureTokenProvider(tenantKey, user);
+
+    // The RouterliciousDocumentServiceFactory creates the driver that allows connections to the Tinylicious service.
+    const documentServiceFactory = new RouterliciousDocumentServiceFactory(
+        tokenProvider,
+    );
+
+    // The code loader provides the ability to load code packages that have been quorumed on and that represent
+    // the code for the document.
+    const codeLoader = getCodeLoaderForPackage(packageId);
+
+    // Finally with all the above objects created we can construct the loader
+    const loader = new Loader({
+        urlResolver,
+        documentServiceFactory,
+        codeLoader,
+    });
+
+    let container: Container;
+    if (shouldCreateNew) {
+        // This flow is used to create a new container and then attach it to the storage.
+        container = await loader.createDetachedContainer({
+            package: packageId.fullId,
+        });
+        await container.attach(urlResolver.createCreateNewRequest(containerId));
+    } else {
+        // This flow is used to get the existing container.
+        container = await loader.resolve({ url });
+    }
+
+    // Wait for connection so that proposals can be sent
+    if (!container.connected) {
+        console.log("waiting for the container to get connected");
+        await new Promise<void>((resolve, reject) => {
+            container?.once("connected", () => resolve());
+            container?.once("closed", (error) => {
+                if (error?.message === "ExistingContextDoesNotSatisfyIncomingProposal") {
+                    reject(
+                        new Error(
+                            `Loaded code is not compatible with the document schema.`,
+                        ),
+                    );
+                } else {
+                    reject(
+                        new Error(
+                            `Container closed unexpectedly. ${error?.message}`,
+                        ),
+                    );
+                }
+            });
+        });
+    }
+
+    // The getFluidObjectAndRender helper method performs the rendering of the data store identified
+    // by the URL in the browser.
+    await getFluidObjectAndRender(
+        loader,
+        container,
+        url,
+        document.getElementById("content") as HTMLDivElement,
+    );
+
+    return container;
+}
+
+if (document.location.pathname === "/") {
+    // Redirect to create a new document with an auto-generated container ID
+    // when URL parameters were not specified by the user
+    const newContainerId = Date.now().toString();
+    window.location.href = `/${newContainerId}?code=${defaultPackage}#CreateNew`;
+} else {
+    const appParams = parseAppUrl();
+    const packageDetails = extractPackageIdentifierDetails(appParams.packageName);
```

Comment:
I also wonder if we should drop the package from the URL entirely.  It could maybe be ok on first load (establishing the first code details to go into the new container), but on all subsequent loads we should definitely be pulling the code details from the container itself rather than from the URL.

For example, proposing a major upgrade and then hitting F5 (rather than clicking the reload button) breaks the demo because the URL doesn't get updated with the latest package version.  If we pulled from the container instead of the URL we could load successfully.

Reply:
I agree, that would make the sample app cleaner and more intuitive. I refactored the app with this proposed approach.